starknet_py/utils/typed_data.py

```diff
@@ -164,12 +168,35 @@ def _encode_data(self, type_name: str, data: dict) -> List[int]:
         return values
 
     def _verify_types(self):
-        reserved_type_names = ["felt", "felt*", "string", "selector", "merkletree"]
+        if self.domain.separator_name not in self.types:
+            raise ValueError(f"Types must contain '{self.domain.separator_name}'.")
+
+        reserved_type_names = ["felt", "string", "selector", "merkletree"]
 
         for type_name in reserved_type_names:
             if type_name in self.types:
                 raise ValueError(f"Reserved type name: {type_name}")
 
+        referenced_types = [
+            parameter for type_name in self.types for parameter in self.types[type_name]
+        ]
+        referenced_types = [
+            ref_type.contains
+            if ref_type.contains is not None
+            else strip_pointer(ref_type.type)
+            for ref_type in referenced_types
+        ] + [self.domain.separator_name, self.primary_type]
+
+        for type_name in self.types:
+            if not type_name:
+                raise ValueError("Type names cannot be empty.")
+            if is_pointer(type_name):
+                raise ValueError(f"Type names cannot end in *. {type_name} was found.")
+            if type_name not in referenced_types:
+                raise ValueError(
+                    f"Dangling types are not allowed. Unreferenced type {type_name} was found."
```

Comment:
Can you explain why it is actually not allowed?

Reply:
It's one of the rules of [types identification](https://github.com/starknet-io/SNIPs/pull/77/files#diff-d594c362a3c260ea39dacaca55a6f02927013cebd19a5c8d202117ff9147ff14R126)
Here's a link to the [subsection about no dangling types](https://github.com/starknet-io/SNIPs/pull/77/files#diff-d594c362a3c260ea39dacaca55a6f02927013cebd19a5c8d202117ff9147ff14R126)